controllers/halayerset_controller.go

```diff
@@ -18,46 +18,603 @@ package controllers
+	service.Namespace = hals.Namespace
+
+	service.Name = serviceName
+	service.Spec.Selector = map[string]string{"app": haClusterName}
+	service.Spec.Type = corev1.ServiceTypeClusterIP
+
+	ports := []corev1.ServicePort{{Protocol: corev1.ProtocolTCP, Port: 2224, Name: "pcs"}}
+	for i := 1; i < 10; i++ {
+		ports = append(ports, corev1.ServicePort{Protocol: corev1.ProtocolUDP, Port: int32(5403 + i), Name: fmt.Sprintf("corosync%v", i)})
+	}
+	service.Spec.Ports = ports
+}
+
+func (r *HALayerSetReconciler) createHALayer(hals *appv1alpha1.HALayerSet) error {
+	if err := r.createHAService(hals); err != nil {
+		return err
+	}
+	if err := r.createHAPod(hals); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *HALayerSetReconciler) deleteHALayer(namespace string) error {
+	if err := r.deleteHAService(namespace); err != nil {
+		return err
+	}
+
+	if err := r.deleteHAPod(namespace); err != nil {
+		return err
+	}
+
+	return nil
+}
+
 // SetupWithManager sets up the controller with the Manager.
 func (r *HALayerSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&appv1alpha1.HALayerSet{}).
 		Complete(r)
 }
+
+func (r *HALayerSetReconciler) verifyDeployment(deploymentName string, namespace string) error {
+	deployment := new(v1.Deployment)
+	key := client.ObjectKey{Name: deploymentName, Namespace: namespace}
+	var err error
+	if err = r.Client.Get(context.Background(), key, deployment); err != nil {
+		if errors.IsNotFound(err) {
+			r.Log.Error(err, "Can't manage deployment because it does not exist", "deployment name", deploymentName)
+		} else {
+			r.Log.Error(err, "Error fetching deployment", "deployment name", deploymentName)
+		}
+		return err
+	}
+	if deployment.Spec.Replicas != nil && *deployment.Spec.Replicas > 0 {
+		r.Log.Error(err, "Managed deployment replicas should be 0 ", "deployment name", deploymentName, "replicas", *deployment.Spec.Replicas)
+		return err
+	}
+	return nil
+}
+
+func (r *HALayerSetReconciler) getHAPod(namespace string) (*corev1.Pod, error) {
+	key := client.ObjectKey{Name: podNameHALayer, Namespace: namespace}
+	pod := new(corev1.Pod)
+	if err := r.Client.Get(context.Background(), key, pod); err != nil {
+		if !errors.IsNotFound(err) {
+			r.Log.Error(err, "failed fetching HA layer pod", "pod name", podNameHALayer)
+		}
+		return nil, err
+	}
+	return pod, nil
+
+}
+
+func (r *HALayerSetReconciler) verifyPodIsRunning(hals *appv1alpha1.HALayerSet) error {
+	var pod *corev1.Pod
+	var err error
+	var podPhase corev1.PodPhase
+	waitTimeLeftSeconds := 30
+	for waitTimeLeftSeconds > 0 {
+		if pod, err = r.getHAPod(hals.Namespace); err != nil {
+			time.Sleep(time.Second)
+			waitTimeLeftSeconds--
+		} else {
+			err = nil
+			podPhase = pod.Status.Phase
+
+			if r.isPodRunning(podPhase) {
+				return nil
+			}
+			time.Sleep(time.Second)
+			waitTimeLeftSeconds--
+		}
+	}
```

Comment:
@n1r1 I'm thinking now of refactoring according to your advice.
Here is my dilemma - I'd like to log an error after X seconds the pod is not running.
If I requeue the request I can't keep track of the time period that has passed (I thought of keeping this data on the CR but it seems like an overkill).
Is there a simple solution I'm overlooking ?

Reply:
looks like `getHAPod()` already handles the case of non-existing pod so maybe you can use the pod's creation timestamp to determine the time that passed since the pod was created

Reply:
It's a nice idea it would have worked if I had different handling for **pod not found**/**pod not running** use cases , but since I treat both cases the same way (log an error after X seconds) it will not solve my **pod not found** use case.

Reply:
I can probably explain better in a voice chat.
let me know when you have few minutes to discuss